### PR TITLE
SONIC updates for site support

### PR DESCRIPTION
**Pull request by kpedro88**
#### PR description:

This PR includes several updates to support production-style tests at several sites, including T2_US_Purdue, other T2s, and NERSC. There are also some important fixes for bugs that were found during the course of testing.

New features/changes:
* A failed response to a repository index query is now interpreted to indicate that the server should be skipped. (The new load balancer developed at Purdue will intercept the repository index query and return an error if it detects that a GPU is saturated.)
* A site can provide a local GPU server using environment variables `$SONIC_LOCAL_BALANCER_HOST` and `$SONIC_LOCAL_BALANCER_PORT`, set via `cmsset_local.sh` in `SITECONF`. (This is a temporary solution that is eventually expected to migrate to something more standardized, similar to `storage.json`. It is introduced here to facilitate tests at other T2 sites to gain operational experience before determining the final form of this configuration.)
* Separate message category called `TritonDiscovery` to enable tracking/debugging of production-style jobs without requiring full verbosity. (This can be seen as an interim step toward full provenance tracking.)
    * Along these lines, improve handling of verbose output options in test script.
* Support is added for `podman` and `podman-hpc` to launch Triton containers (useful at NERSC).
* Failures of optional server requests are no longer treated as exceptions, to avoid unnecessary job failures.

Bug fixes:
* Modify some apptainer workarounds (remove an old one, add a new one...)
* Check for the presence of a GPU in the `cmsTriton` server launching script, rather than in Python. (Production-style jobs can dump the Python config on a local node and then reuse it on a worker node, so any system calls happen on the local node and may not give the right answer for the worker node as to the presence of a GPU.)
* Fix a bug in the implementation of thread controls for models on the CPU fallback server.

There are some interface changes (related to supporting more container engines and checking for local GPUs), which are documented.

#### PR validation:

Unit tests succeed. Site-specific options have been tested successfully at the appropriate sites.

#### If this PR is a backport please specify the original PR and why you need to backport that PR. If this PR will be backported please specify to which release cycle the backport is meant for:

Not a backport and not intended to be backported.

### CODE REVIEW

**Comment by kpedro88** [comment]
test parameters:
workflows = 10805.31,11634.9001,24834.9001
relvals_opt = --what cleanedupgrade,standard,highstats,pileup,generator,extendedgen,production,identity,ged,machine,premix,nano,gpu,2017,2026

**Comment by kpedro88** [comment]
please test

**Comment by kpedro88** [comment]
please test

**Comment by kpedro88** [comment]
please test

**Comment by kpedro88** [comment]
please test

**Comment by fwyzard** [comment]
To clarify the reason for the lack of comments: I am not the best person to review SONIC-related changes, and @makortel has looked at these PRs in the past.
He is on vacation, and should be back the week of July 6th.

**Comment by kpedro88** [comment]
@makortel a gentle ping to review this PR when you have the chance :)

**Comment by makortel** [comment]
I'm very close in getting to it :)

**Comment by kpedro88** [comment]
Updated to address a segmentation violation that was recently found, as well as to address #45454.

**Comment by makortel** [comment]
@cmsbuild, please test

File: HeterogeneousCore/SonicTriton/src/TritonService.cc [hunk 1]
```diff
@@ -286,16 +322,35 @@ void TritonService::preBeginJob(edm::PathsAndConsumesOfModulesBase const&, edm::
...
+      if (chosenDevice == "cpu")
+        server.type = TritonServerType::LocalCPU;
+      else if (chosenDevice == "gpu")
+        server.type = TritonServerType::LocalGPU;
+      else
+        throw cms::Exception("FallbackFailed")
+            << "TritonService: unsupported device choice " << chosenDevice << " for fallback server, log follows:\n"
+            << output;
+    } else
+      throw cms::Exception("FallbackFailed")
```
Comment (makortel) [inline HeterogeneousCore/SonicTriton/src/TritonService.cc:340]: I feel these exceptions would be better as
```suggestion
      throw edm::Exception(edm::errors::ExternalFailure)
```
(following https://github.com/cms-sw/cmssw/pull/43274)

File: HeterogeneousCore/SonicTriton/src/TritonService.cc [hunk 2]
```diff
@@ -286,16 +322,35 @@ void TritonService::preBeginJob(edm::PathsAndConsumesOfModulesBase const&, edm::
...
-  if (pos != std::string::npos) {
-    auto pos2 = pos + portIndicator.size();
-    auto pos3 = output.find('\n', pos2);
-    const auto& portNum = output.substr(pos2, pos3 - pos2);
+  const auto& portNum = extractFromLog(output, "CMS_TRITON_GRPC_PORT: ");
+  if (!portNum.empty())
     server.url += ":" + portNum;
-  } else
+  else
     throw cms::Exception("FallbackFailed") << "TritonService: Unknown port for fallback server, log follows:\n"
```
Comment (makortel) [inline HeterogeneousCore/SonicTriton/src/TritonService.cc:354]: I'd then use `throw edm::Exception(edm::errors::ExternalFailure)` here too

File: HeterogeneousCore/SonicTriton/src/TritonService.cc [hunk 3]
```diff
@@ -74,34 +91,38 @@ TritonService::TritonService(const edm::ParameterSet& pset, edm::ActivityRegistr
...
-      serverType = TritonServerType::LocalGPU;
-#endif
-
-    servers_.emplace(std::piecewise_construct,
-                     std::forward_as_tuple(Server::fallbackName),
-                     std::forward_as_tuple(Server::fallbackName, Server::fallbackAddress, serverType));
-  }
+  //check for server specified in SITECONF
+  std::string siteconf_address(edm::getEnvironmentVariable(Server::siteconfName + "_HOST"));
+  std::string siteconf_port(edm::getEnvironmentVariable(Server::siteconfName + "_PORT"));
```
Comment (makortel) [inline HeterogeneousCore/SonicTriton/src/TritonService.cc:96]: I'd add here a comment that the environment variables are a temporary solution, eventually to be replaced with site local config.

**Comment by kpedro88** [comment]
please test

**Comment by makortel** [comment]
Comparison failures are related to https://github.com/cms-sw/cmssw/issues/39803 and https://github.com/cms-sw/cmssw/issues/45505

**Comment by makortel** [comment]
+heterogeneous

**Comment by kpedro88** [comment]
@cms-sw/reconstruction-l2 please check - the only change in your area is propagating some interface changes to a unit test

**Comment by makortel** [comment]
@kpedro88 Could you backport the slc7 workaround (https://github.com/cms-sw/cmssw/pull/45182/commits/516de1d7d8d3ae696ce3fb64bf102d7e221e1899) to 14_0_X? ORP would like to get the slc7 unit test failures (https://github.com/cms-sw/cmssw/issues/45454) fixed there too.

**Comment by mandrenguyen** [comment]
+reconstruction
@kpedro88 Are the comments from Matti still relevant?  If not can you resolve them?

**Comment by kpedro88** [comment]
@mandrenguyen the comments were resolved in https://github.com/cms-sw/cmssw/pull/45182/commits/e24f36272789d033f0196e050c206aa50766c0ad, now marked as such.

**Comment by mandrenguyen** [comment]
+1

**Comment by makortel** [comment]
> @makortel the use of OSG apptainer was introduced in 14_1 (#43814). Any failures in 14_0 are therefore related to other apptainer incompatibilities. I will make a backport PR with the relevant commits.

Ah right, the error in 14_0_X is indeed different (
```
INFO:    Disabling cgroups because systemd is unavailable
ERROR:   container cleanup failed: no instance found with name triton_server_instance_CPU_4cb5181e-0d20-4231-b181-0e53d40c7b4d
FATAL:   container creation failed: while applying cgroups config: while creating cgroup manager: systemd not running on this host, cannot use systemd cgroups manager

FATAL:   while executing starter: failed to start instance: while running /usr/libexec/apptainer/bin/starter: exit status 255
```
https://cmssdt.cern.ch/SDT/cgi-bin/logreader/slc7_amd64_gcc12/CMSSW_14_0_MULTIARCHS_X_2024-08-05-2300/unitTestLogs/HeterogeneousCore/SonicTriton#/74-74 )

Thanks!

**Comment by kpedro88** [comment]
@makortel see #45652 and #45653 (the workaround is actually now needed in this release cycle also... it must be a relatively new problem that happened to pop up after I last ran the unit tests locally when testing this PR)

**Comment by makortel** [comment]
Thanks!

**Pull request merged**